### ISSUE TITLE
Moving average ignored a value of inf/nan

### ISSUE THREAD

**Issue by emjay73** [dbolya/yolact#222]
Hi, me again.
"Moving average ignored a value of inf/nan "
issue happens again with multi-GPUs!

Would you please take a look at this issue once again?

![Screenshot from 2019-11-22 21-26-17](https://user-images.githubusercontent.com/36818801/69425846-0b47b080-0d6f-11ea-8b90-ce78b9fe193b.png)

(prev issue link : https://github.com/dbolya/yolact/issues/186)
Thanks!

**Comment by dbolya** [comment]
What was your command used to run this? And are you on the latest master?

**Comment by emjay73** [comment]
Yes, it was the latest one.
But to make it clear, I just cloned the repository to newly created folder and changed the coco DB location in the config file so that I can use my own coco DB without moving it to the 'data' folder.
After that, I did the following.
export CUDA_VISIBLE_DEVICES=0,1,2,3
python train.py --config=yolact_base_config --batch_size=32

and it failed once again.

![Screenshot from 2019-11-25 11-23-21](https://user-images.githubusercontent.com/36818801/69507526-32bc9a00-0f76-11ea-8a3f-ce3e67370dc4.png)

![Screenshot from 2019-11-25 11-23-52](https://user-images.githubusercontent.com/36818801/69507533-38b27b00-0f76-11ea-98a7-3e6ca0cc98c1.png)

![Screenshot from 2019-11-25 11-25-50](https://user-images.githubusercontent.com/36818801/69507577-5aabfd80-0f76-11ea-8d69-1a3f7cd0270d.png)

**Comment by dbolya** [comment]
Wow that's quite weird. This is the unmodified base config? I'll see if I can reproduce this on monday.

And just to sanity check, does it happen with 1 gpu?

**Comment by emjay73** [comment]
As I mentioned above, I changed the coco DB location in the config file so that I can use my coco DB without moving it to the 'data' folder.

![Screenshot from 2019-11-25 11-45-03](https://user-images.githubusercontent.com/36818801/69508438-1110e200-0f79-11ea-9794-7a132f1b16f6.png)

Except that, the other values have the default values.
I started training with the following command.

export CUDA_VISIBLE_DEVICES=0
python train.py --config=yolact_base_config --batch_size=8

I'll let you know what happens.

**Comment by emjay73** [comment]
Quick related question!
When do you suppose it is trained enough?

**Comment by dbolya** [comment]
You mean as to what point it the loss shouldn't be able to explode anymore? That should be around epoch 1-2, but evidently it can still explode on epoch 2 as you found. I'm currently testing 4 GPUs on yolact_base_config, I'll let you know if I can reproduce this.

**Comment by emjay73** [comment]
I trained it roughly for a day and it's safe!

![Screenshot from 2019-11-26 09-59-58](https://user-images.githubusercontent.com/36818801/69590683-96f16380-1033-11ea-952b-2bc85369eea0.png)

**Comment by emjay73** [comment]
With second attempt, loss exploded with single GPU :<
![Screenshot from 2019-11-26 17-17-18](https://user-images.githubusercontent.com/36818801/69611264-bd81bf80-1070-11ea-83de-f82df86edc35.png)
![Screenshot from 2019-11-26 17-17-33](https://user-images.githubusercontent.com/36818801/69611269-c2df0a00-1070-11ea-8da7-e9b7ee233571.png)
![Screenshot from 2019-11-26 17-17-40](https://user-images.githubusercontent.com/36818801/69611279-c8d4eb00-1070-11ea-9c43-4375d29b7a96.png)

**Comment by dbolya** [comment]
Ok this is an issue, I was able to reproduce this on 4 gpus. I'm currently testing on 1 gpu, but dang everything was working fine with pascal, no idea why it's not working with COCO.

I'll take a deeper look as to why this is happening and see if I can fix it.

**Comment by emjay73** [comment]
I hope all of those problems that occurred in single/multi GPU cases are originated from the same flaw.
Thanks and  plz,
You are my only hope T-T

**Comment by emjay73** [comment]
And what about gradient clipping?

**Comment by dbolya** [comment]
Ok sorry about the delay (thanksgiving and all). The single GPU test came back successful and reproduced the mAP in the paper, so I guess the issue may be somewhere else? Now that I'm back, I'll investigate this further.

**Comment by emjay73** [comment]
OK Thanks~

**Comment by dbolya** [comment]
An update: I figured out the source of the random infs during training that you see every now and again. Augmentation could actually output detections with *0 bbox area*, hence why some things are divided by 0 and thus produce infinity. Since I ignore batches with inf loss, this wouldn't be the cause of the explosion, but it might be boxes with really small area instead that's causing this explosion (since I don't filter those out). 

Right now I'm trying out discarding any box with fewer than than 4 pixels of width and 4 pixels of height. I haven't seen a single inf or nan so far, so this looks promising. I'll update you once / if it finishes training.

**Comment by emjay73** [comment]
That's good news!
And well, it seems like the problem that is not related to the multiple GPUs..?

**Comment by HuangLian126** [comment]
@dbolya Thanks for your work. Do you finish the training that discarding any box with fewer than than 4 pixels of width and 4 pixels of height ?

**Comment by dbolya** [comment]
It's epoch 49/54 but that's enough for me to verify that it reached the same mAP as in the paper with 4 GPUs and batch size 28 (so 7 / gpu) with no loss explosion! I'll push the code now so you guys can test.

**Comment by dbolya** [comment]
Pushed the change. Let me know if I borked anything!

**Comment by emjay73** [comment]
It works fine at epoch 45/54.
Thank you!
